services/spar/src/Spar/Scim/User.hs

```diff
@@ -61,31 +64,101 @@ import qualified Web.Scim.Schema.ResourceType     as Scim
+  -- getUsers itself will not do any mutations in the SCIM database; as it
+  -- feels weird to me (arian) that a GET causes side-effects like that.
+  -- Instead, 'updateUser' will when updating a user see if the user was
+  -- already in the SCIM Table; and if not insert it on demand, just like
+  -- 'POST'.
   getUsers :: ScimTokenInfo
            -> Maybe Scim.Filter
            -> Scim.ScimHandler Spar (Scim.ListResponse (Scim.StoredUser SparTag))
-  getUsers ScimTokenInfo{stiTeam} mbFilter = do
-    members <- lift $ getTeamMembers stiTeam
+  getUsers ScimTokenInfo{stiTeam} Nothing = do
+    -- TODO(arianvp): DEPRECATED
+    -- Getting users without a filter is deprecated an must be
+    -- removed in the next release. It is never what we want; and non of the
+    -- current SCIM implementations mandate its use. 
+    --
+    -- The code is rather problematic as it does O(n) HTTP requests to Brig (as
+    -- brig does not support getting users in bulk). Given that Azure would
+    -- send a /Users request _per user_ this would yield in O(n^2) requests,
+    -- which blows up extremely quickly for large teams.
+    --
+    -- NOTE: unless you supply a filter, it will only list
+    -- SCIM users, not all users in a team. That was the original behaviour
+    -- and we didn't feel like changing it for this as it's going to be removed
+    -- soon anyway. Please scream loudly in the review if you disagree with this,
+    -- but it would complicate this code a bit, instead of leaving it as is.
+    members <- lift $ Galley.getTeamMembers stiTeam
     brigusers :: [User]
       <- filter (not . userDeleted) <$>
          lift (Intra.Brig.getBrigUsers ((^. Galley.userId) <$> members))
     scimusers :: [Scim.StoredUser SparTag]
       <- lift . wrapMonadClient . Data.getScimUsers $ BrigTypes.userId <$> brigusers
-    let check user = case mbFilter of
-          Nothing -> pure True
-          Just filter_ ->
-            let user' = Scim.value (Scim.thing user)
-            in case Scim.filterUser filter_ user' of
-                 Right res -> pure res
-                 Left err  -> throwError $
-                   Scim.badRequest Scim.InvalidFilter (Just err)
-    -- FUTUREWORK: once bigger teams arrive, we should have pagination here.
-    Scim.fromList <$> filterM check scimusers
+    pure $ Scim.fromList scimusers
+
+  getUsers ScimTokenInfo{stiTeam, stiIdP} (Just filter') = do
+    idp <- stiIdP ?? Scim.serverError "No IdP configured for the provisioning token"
+    idpConfig <- (wrapMonadClient . Data.getIdPConfig $ idp)  !? Scim.serverError "No IdP configured for the provisioning token"
+    case filter' of
+      Scim.FilterAttrCompare (Scim.AttrPath schema attrName _subAttr) Scim.OpEq (Scim.ValString val)
+        | Scim.isUserSchema schema -> do
+            let
+              createOrGetScimUser :: BrigTypes.User -> MaybeT (Scim.ScimHandler Spar) (Scim.StoredUser SparTag)
```

Comment:
Could we move this to a separate function or a where clause? I find it hard to read this `case` statement which is really about checking the `attribute`?

Reply:
Also: wouldn't it make this function simpler also if we don't use `MaybeT` but rather `Spar` directly?

Reply:
We need MaybeT for the short-circuit behaviour instead of returning an error. It's only used in this specific function. E.g. lists should never return 404 but an empty list instead